Fix building with tss-esapi 7.0+

It seems that there's been changes with the newer rust tss-esapi 7.x release:

```
   Compiling tpm2-policy v0.5.3 (/builddir/build/BUILD/tpm2-policy-0.5.3)
error[E0432]: unresolved import `tss_esapi::utils::AsymSchemeUnion`
  --> src/lib.rs:26:5
   |
26 |     utils::AsymSchemeUnion,
   |     ^^^^^^^^^^^^^^^^^^^^^^ no `AsymSchemeUnion` in `utils`
error[E0432]: unresolved import `tss_esapi::utils::AsymSchemeUnion`
  --> src/structures.rs:22:5
   |
22 |     utils::AsymSchemeUnion,
   |     ^^^^^^^^^^^^^^^^^^^^^^ no `AsymSchemeUnion` in `utils`
error[E0433]: failed to resolve: could not find `Tpm2BPublicBuilder` in `utils`
   --> src/structures.rs:191:38
    |
191 |                 Ok(tss_esapi::utils::Tpm2BPublicBuilder::new()
    |                                      ^^^^^^^^^^^^^^^^^^ could not find `Tpm2BPublicBuilder` in `utils`
error[E0433]: failed to resolve: could not find `PublicParmsUnion` in `utils`
   --> src/structures.rs:194:51
    |
194 |                     .with_parms(tss_esapi::utils::PublicParmsUnion::RsaDetail(
    |                                                   ^^^^^^^^^^^^^^^^ could not find `PublicParmsUnion` in `utils`
error[E0433]: failed to resolve: could not find `TpmsRsaParmsBuilder` in `utils`
   --> src/structures.rs:195:43
    |
195 |                         tss_esapi::utils::TpmsRsaParmsBuilder::new_unrestricted_signing_key(
    |                                           ^^^^^^^^^^^^^^^^^^^ could not find `TpmsRsaParmsBuilder` in `utils`
error[E0433]: failed to resolve: could not find `PublicIdUnion` in `utils`
   --> src/structures.rs:203:52
    |
203 |                     .with_unique(tss_esapi::utils::PublicIdUnion::Rsa(rsa_uniq))
    |                                                    ^^^^^^^^^^^^^ could not find `PublicIdUnion` in `utils`
error[E0433]: failed to resolve: could not find `SignatureData` in `utils`
   --> src/lib.rs:239:46
    |
239 |                 signature: tss_esapi::utils::SignatureData::RsaSignature(policy.signature.clone()),
    |                                              ^^^^^^^^^^^^^ could not find `SignatureData` in `utils`
error[E0412]: cannot find type `AsymSchemeUnion` in module `tss_esapi::utils`
   --> src/structures.rs:150:66
    |
150 |     pub(crate) fn get_signing_scheme(&self) -> tss_esapi::utils::AsymSchemeUnion {
    |                                                                  ^^^^^^^^^^^^^^^ not found in `tss_esapi::utils`
error[E0422]: cannot find struct, variant or union type `Signature` in module `tss_esapi::utils`
   --> src/lib.rs:237:47
    |
237 |             let signature = tss_esapi::utils::Signature {
    |                                               ^^^^^^^^^ not found in `tss_esapi::utils`
    |
help: consider importing this variant
    |
15  | use tss_esapi::constants::Tss2ResponseCodeKind::Signature;
    |
error[E0308]: mismatched types
   --> src/lib.rs:135:76
    |
135 |                     ctx.execute_without_session(|context| context.pcr_read(&pcr_sel))?;
    |                                                                            ^^^^^^^^ expected struct `PcrSelectionList`, found reference
    |
    = note: expected struct `PcrSelectionList`
            found reference `&std::result::Result<PcrSelectionList, tss_esapi::Error>`
error[E0599]: no method named `pcr_bank` found for struct `DigestList` in the current scope
   --> src/lib.rs:141:30
    |
141 | ...                   .pcr_bank(pcr_hash_alg)
    |                        ^^^^^^^^ method not found in `DigestList`
error[E0308]: mismatched types
   --> src/lib.rs:153:25
    |
153 |                         &concatenated_pcr_values,
    |                         ^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `MaxBuffer`, found `&MaxBuffer`
    |
help: consider removing the borrow
    |
153 -                         &concatenated_pcr_values,
153 +                         concatenated_pcr_values,
    | 
error[E0308]: mismatched types
   --> src/lib.rs:159:60
    |
159 |                 ctx.policy_pcr(policy_session.try_into()?, &hashed_data, pcr_sel)?;
    |                                                            ^^^^^^^^^^^^ expected struct `Digest`, found `&Digest`
    |
help: consider removing the borrow
    |
159 -                 ctx.policy_pcr(policy_session.try_into()?, &hashed_data, pcr_sel)?;
159 +                 ctx.policy_pcr(policy_session.try_into()?, hashed_data, pcr_sel)?;
    | 
error[E0308]: mismatched types
   --> src/lib.rs:172:59
    |
172 |                 let loaded_key = ctx.load_external_public(&tpm_signkey, Hierarchy::Owner)?;
    |                                                           ^^^^^^^^^^^^ expected enum `Public`, found `&TPM2B_PUBLIC`
error[E0308]: mismatched types
   --> src/lib.rs:199:21
    |
199 |                     &approved_policy,
    |                     ^^^^^^^^^^^^^^^^ expected struct `Digest`, found `&Digest`
    |
help: consider removing the borrow
    |
199 -                     &approved_policy,
199 +                     approved_policy,
    | 
error[E0308]: mismatched types
   --> src/lib.rs:200:21
    |
200 |                     &policy_ref,
    |                     ^^^^^^^^^^^ expected struct `tss_esapi::structures::Nonce`, found `&tss_esapi::structures::Nonce`
    |
help: consider removing the borrow
    |
200 -                     &policy_ref,
200 +                     policy_ref,
    | 
error[E0308]: mismatched types
   --> src/lib.rs:235:23
    |
235 |                 .hash(&ahash, HashingAlgorithm::Sha256, Hierarchy::Null)?
    |                       ^^^^^^ expected struct `MaxBuffer`, found `&MaxBuffer`
    |
help: consider removing the borrow
    |
235 -                 .hash(&ahash, HashingAlgorithm::Sha256, Hierarchy::Null)?
235 +                 .hash(ahash, HashingAlgorithm::Sha256, Hierarchy::Null)?
    | 
error[E0308]: mismatched types
   --> src/lib.rs:241:56
    |
241 |             let tkt = ctx.verify_signature(loaded_key, &ahash, signature)?;
    |                                                        ^^^^^^ expected struct `Digest`, found `&Digest`
    |
help: consider removing the borrow
    |
241 -             let tkt = ctx.verify_signature(loaded_key, &ahash, signature)?;
241 +             let tkt = ctx.verify_signature(loaded_key, ahash, signature)?;
    | 
Some errors have detailed explanations: E0308, E0412, E0422, E0432, E0433, E0599.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `tpm2-policy` due to 18 previous errors
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -10,7 +10,7 @@ license = "EUPL-1.2"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-tss-esapi = "6.1.1"
+tss-esapi = "7.0.0"
 serde = "1.0"
 base64 = "0.12.1"
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -21,9 +21,10 @@ use tss_esapi::{
     handles::KeyHandle,
     interface_types::algorithm::HashingAlgorithm,
     interface_types::{resource_handles::Hierarchy, session_handles::AuthSession},
-    structures::SymmetricDefinition,
-    structures::{Digest, MaxBuffer, Nonce, PcrSelectionListBuilder, PcrSlot, VerifiedTicket},
-    utils::AsymSchemeUnion,
+    structures::{
+        Digest, MaxBuffer, Nonce, PcrSelectionListBuilder, PcrSlot, SignatureScheme, VerifiedTicket,
+    },
+    structures::{Signature, SymmetricDefinition},
 };
 
 mod error;
@@ -129,34 +130,28 @@ impl TPMPolicyStep {
 
                 let pcr_sel = PcrSelectionListBuilder::new()
                     .with_selection(pcr_hash_alg, &pcr_ids)
-                    .build();
+                    .build()?;
 
                 let (_update_counter, pcr_sel, pcr_data) =
-                    ctx.execute_without_session(|context| context.pcr_read(&pcr_sel))?;
+                    ctx.execute_without_session(|context| context.pcr_read(pcr_sel))?;
 
-                let concatenated_pcr_values: Result<Vec<&[u8]>> = pcr_ids
+                let concatenated_pcr_values = pcr_data
+                    .value()
                     .iter()
-                    .map(|x| {
-                        Ok(pcr_data
-                            .pcr_bank(pcr_hash_alg)
-                            .ok_or_else(|| Error::PcrValueNotReturned(pcr_hash_alg, None))?
-                            .pcr_value(*x)
-                            .ok_or_else(|| Error::PcrValueNotReturned(pcr_hash_alg, Some(*x)))?
-                            .value())
-                    })
-                    .collect();
-                let concatenated_pcr_values = concatenated_pcr_values?.as_slice().concat();
+                    .map(|x| x.value())
+                    .collect::<Vec<&[u8]>>()
+                    .concat();
                 let concatenated_pcr_values = MaxBuffer::try_from(concatenated_pcr_values)?;
 
                 let (hashed_data, _ticket) = ctx.execute_without_session(|context| {
                     context.hash(
-                        &concatenated_pcr_values,
+                        concatenated_pcr_values,
                         HashingAlgorithm::Sha256,
                         Hierarchy::Owner,
                     )
                 })?;
 
-                ctx.policy_pcr(policy_session.try_into()?, &hashed_data, pcr_sel)?;
+                ctx.policy_pcr(policy_session.try_into()?, hashed_data, pcr_sel)?;
                 next._send_policy(ctx, policy_session)
             }
 
@@ -168,8 +163,8 @@ impl TPMPolicyStep {
             } => {
                 let policy_ref = Nonce::try_from(policy_ref)?;
 
-                let tpm_signkey = tss_esapi::tss2_esys::TPM2B_PUBLIC::try_from(&signkey)?;
-                let loaded_key = ctx.load_external_public(&tpm_signkey, Hierarchy::Owner)?;
+                let loaded_key =
+                    ctx.load_external_public(signkey.clone().try_into()?, Hierarchy::Owner)?;
                 let loaded_key_name = ctx.tr_get_name(loaded_key.into())?;
 
                 let (approved_policy, check_ticket) = match policies {
@@ -196,8 +191,8 @@ impl TPMPolicyStep {
 
                 ctx.policy_authorize(
                     policy_session.try_into()?,
-                    &approved_policy,
-                    &policy_ref,
+                    approved_policy,
+                    policy_ref,
                     &loaded_key_name,
                     check_ticket,
                 )?;
@@ -215,7 +210,7 @@ fn find_and_play_applicable_policy(
     policies: &[SignedPolicy],
     policy_session: AuthSession,
     policy_ref: &[u8],
-    scheme: AsymSchemeUnion,
+    scheme: SignatureScheme,
     loaded_key: KeyHandle,
 ) -> Result<(Digest, VerifiedTicket)> {
     for policy in policies {
@@ -232,13 +227,24 @@ fn find_and_play_applicable_policy(
             let ahash = MaxBuffer::try_from(ahash)?;
 
             let ahash = ctx
-                .hash(&ahash, HashingAlgorithm::Sha256, Hierarchy::Null)?
+                .hash(ahash, HashingAlgorithm::Sha256, Hierarchy::Null)?
                 .0;
-            let signature = tss_esapi::utils::Signature {
-                scheme,
-                signature: tss_esapi::utils::SignatureData::RsaSignature(policy.signature.clone()),
+            let signature = match scheme {
+                SignatureScheme::RsaPss { hash_scheme } => {
+                    Signature::RsaPss(tss_esapi::structures::RsaSignature::create(
+                        hash_scheme.hashing_algorithm(),
+                        policy.signature.clone().try_into()?,
+                    )?)
+                }
+                SignatureScheme::RsaSsa { hash_scheme } => {
+                    Signature::RsaSsa(tss_esapi::structures::RsaSignature::create(
+                        hash_scheme.hashing_algorithm(),
+                        policy.signature.clone().try_into()?,
+                    )?)
+                }
+                _ => todo!(),
             };
-            let tkt = ctx.verify_signature(loaded_key, &ahash, signature)?;
+            let tkt = ctx.verify_signature(loaded_key, ahash, signature)?;
 
             return Ok((policy_digest, tkt));
         }

--- a/src/structures.rs
+++ b/src/structures.rs
@@ -14,13 +14,10 @@
 
 use crate::error::{Error, Result};
 
-use std::convert::TryFrom;
+use std::convert::{TryFrom, TryInto};
 
 use serde::{Deserialize, Serialize};
-use tss_esapi::{
-    constants::tss as tss_constants, interface_types::algorithm::HashingAlgorithm,
-    utils::AsymSchemeUnion,
-};
+use tss_esapi::{interface_types::algorithm::HashingAlgorithm, structures::SignatureScheme};
 
 fn serialize_as_base64<S>(bytes: &[u8], serializer: S) -> std::result::Result<S::Ok, S::Error>
 where
@@ -37,7 +34,7 @@ where
         .and_then(|string| base64::decode(&string).map_err(serde::de::Error::custom))
 }
 
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Debug, Serialize, Deserialize, Clone)]
 pub enum SignedPolicyStep {
     PCRs {
         pcr_ids: Vec<u16>,
@@ -50,7 +47,7 @@ pub enum SignedPolicyStep {
     },
 }
 
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Debug, Serialize, Deserialize, Clone)]
 pub struct SignedPolicy {
     // policy_ref contains the policy_ref used in the aHash, used to determine the policy to use from a list
     #[serde(
@@ -70,7 +67,7 @@ pub struct SignedPolicy {
 
 pub type SignedPolicyList = Vec<SignedPolicy>;
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub enum TPMPolicyStep {
     NoStep,
     PCRs(HashingAlgorithm, Vec<u64>, Box<TPMPolicyStep>),
@@ -83,22 +80,37 @@ pub enum TPMPolicyStep {
     Or([Box<TPMPolicyStep>; 8]),
 }
 
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Debug, Serialize, Deserialize, Clone, Copy)]
 pub enum RSAPublicKeyScheme {
     RSAPSS,
     RSASSA,
 }
 
 impl RSAPublicKeyScheme {
-    fn to_scheme(&self, hash_algo: &HashAlgo) -> AsymSchemeUnion {
+    fn to_scheme(&self, hash_algo: HashAlgo) -> SignatureScheme {
         match self {
-            RSAPublicKeyScheme::RSAPSS => AsymSchemeUnion::RSAPSS(hash_algo.into()),
-            RSAPublicKeyScheme::RSASSA => AsymSchemeUnion::RSASSA(hash_algo.into()),
+            RSAPublicKeyScheme::RSAPSS => SignatureScheme::RsaPss {
+                hash_scheme: hash_algo.into(),
+            },
+            RSAPublicKeyScheme::RSASSA => SignatureScheme::RsaSsa {
+                hash_scheme: hash_algo.into(),
+            },
+        }
+    }
+
+    fn to_rsa_scheme(&self, hash_algo: HashAlgo) -> Result<tss_esapi::structures::RsaScheme> {
+        match self {
+            RSAPublicKeyScheme::RSAPSS => {
+                Ok(tss_esapi::structures::RsaScheme::RsaPss(hash_algo.into()))
+            }
+            RSAPublicKeyScheme::RSASSA => {
+                Ok(tss_esapi::structures::RsaScheme::RsaSsa(hash_algo.into()))
+            }
         }
     }
 }
 
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Debug, Serialize, Deserialize, Clone, Copy)]
 pub enum HashAlgo {
     SHA1,
     SHA256,
@@ -110,15 +122,8 @@ pub enum HashAlgo {
     SHA3_512,
 }
 
-impl HashAlgo {
-    fn to_tpmi_alg_hash(&self) -> tss_esapi::tss2_esys::TPMI_ALG_HASH {
-        let alg: HashingAlgorithm = self.into();
-        alg.into()
-    }
-}
-
-impl From<&HashAlgo> for HashingAlgorithm {
-    fn from(halg: &HashAlgo) -> Self {
+impl From<HashAlgo> for HashingAlgorithm {
+    fn from(halg: HashAlgo) -> Self {
         match halg {
             HashAlgo::SHA1 => HashingAlgorithm::Sha1,
             HashAlgo::SHA256 => HashingAlgorithm::Sha256,
@@ -132,7 +137,13 @@ impl From<&HashAlgo> for HashingAlgorithm {
     }
 }
 
-#[derive(Debug, Serialize, Deserialize)]
+impl From<HashAlgo> for tss_esapi::structures::HashScheme {
+    fn from(halg: HashAlgo) -> Self {
+        tss_esapi::structures::HashScheme::new(halg.into())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
 pub enum PublicKey {
     RSA {
         scheme: RSAPublicKeyScheme,
@@ -147,22 +158,22 @@ pub enum PublicKey {
 }
 
 impl PublicKey {
-    pub(crate) fn get_signing_scheme(&self) -> tss_esapi::utils::AsymSchemeUnion {
+    pub(crate) fn get_signing_scheme(&self) -> SignatureScheme {
         match self {
             PublicKey::RSA {
                 scheme,
                 hashing_algo,
                 exponent: _,
                 modulus: _,
-            } => scheme.to_scheme(hashing_algo),
+            } => scheme.to_scheme(*hashing_algo),
         }
     }
 }
 
-impl TryFrom<&PublicKey> for tss_esapi::tss2_esys::TPM2B_PUBLIC {
+impl TryFrom<PublicKey> for tss_esapi::structures::Public {
     type Error = Error;
 
-    fn try_from(publickey: &PublicKey) -> Result<Self> {
+    fn try_from(publickey: PublicKey) -> Result<Self> {
         match publickey {
             PublicKey::RSA {
                 scheme,
@@ -180,27 +191,18 @@ impl TryFrom<&PublicKey> for tss_esapi::tss2_esys::TPM2B_PUBLIC {
                         .with_sign_encrypt(true)
                         .with_restricted(false);
 
-                let len = modulus.len();
-                let mut buffer = [0_u8; 512];
-                buffer[..len].clone_from_slice(&modulus[..len]);
-                let rsa_uniq = Box::new(tss_esapi::tss2_esys::TPM2B_PUBLIC_KEY_RSA {
-                    size: len as u16,
-                    buffer,
-                });
-
-                Ok(tss_esapi::utils::Tpm2BPublicBuilder::new()
-                    .with_type(tss_constants::TPM2_ALG_RSA)
-                    .with_name_alg(hashing_algo.to_tpmi_alg_hash())
-                    .with_parms(tss_esapi::utils::PublicParmsUnion::RsaDetail(
-                        tss_esapi::utils::TpmsRsaParmsBuilder::new_unrestricted_signing_key(
-                            scheme.to_scheme(&hashing_algo),
-                            (modulus.len() * 8) as u16,
-                            *exponent,
+                Ok(tss_esapi::structures::PublicBuilder::new()
+                    .with_public_algorithm(tss_esapi::interface_types::algorithm::PublicAlgorithm::Rsa)
+                    .with_name_hashing_algorithm(hashing_algo.into())
+                    .with_rsa_parameters(tss_esapi::structures::PublicRsaParametersBuilder::new_unrestricted_signing_key(
+                            scheme.to_rsa_scheme(hashing_algo)?,
+                            ((modulus.len() * 8) as u16).try_into()?,
+                            tss_esapi::structures::RsaExponent::create(exponent)?,
                         )
-                        .build()?,
-                    ))
+                        .build()?
+                    )
                     .with_object_attributes(object_attributes.build()?)
-                    .with_unique(tss_esapi::utils::PublicIdUnion::Rsa(rsa_uniq))
+                    .with_rsa_unique_identifier(modulus.clone().try_into()?)
                     .build()?)
             }
         }

--- a/tests/utils.rs
+++ b/tests/utils.rs
@@ -28,7 +28,7 @@ pub fn get_tpm2_ctx() -> Context {
     };
 
     let tcti = Tcti::from_str(&tcti_path).unwrap();
-    unsafe { Context::new(tcti) }.unwrap()
+    Context::new(tcti).unwrap()
 }
 
 fn get_signtool_location() -> PathBuf {

